### ISSUE TITLE
Issue with external imports and commonjs

### ISSUE THREAD

**Issue by charlag** [PepsRyuu/nollup#169]
Hi! Sorry but it's me again :laughing: 

There's an issue with `@rollup/plugin-commonjs` when it lefts synthentic imports like `fs?commonjs-require` in the output. This does not occur with rollup. Here's a branch to demonstrate: https://github.com/tutao/nollup-bug/tree/commonjs
The idea is that you have external import inside the commonjs module (which is the case for many 3rd party dependencies which depend on node or electron).
I think I know where it happens:

https://github.com/rollup/plugins/blob/35cea58e504617d5278b452df9e76e87a416db4b/packages/commonjs/src/generate-imports.js#L148-L150

but I don't understand yet why it does work with rollup. It seems like emitted file must be analyzed again but it's not happening?

EDIT: to clarify, rollup resolves those synthetic imports and they are not present but nollup doesn't and left in the output.

**Comment by PepsRyuu** [comment]
Using the fix you've proposed in the other issue. I'm getting this:

```
Generating...
(node:3448) UnhandledPromiseRejectionWarning: Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'fs?commonjs-require' imported from /home/peps/nollup-bug/build/util1-[hash].js
    at packageResolve (internal/modules/esm/resolve.js:655:9)
    at moduleResolve (internal/modules/esm/resolve.js:696:18)
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:810:11)
    at Loader.resolve (internal/modules/esm/loader.js:85:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:229:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:51:40)
    at link (internal/modules/esm/module_job.js:50:36)
```
Is that right?

**Comment by charlag** [comment]
I think it's the same problem? with that workaround it fails with:

```
(node:9050) UnhandledPromiseRejectionWarning: Error [ERR_MODULE_NOT_FOUND]: Cannot find package ' fs?commonjs-require' imported from /home/ivk/dev/nollup-bug/build/util1-[hash].js
    at packageResolve (internal/modules/esm/resolve.js:674:9)
    at moduleResolve (internal/modules/esm/resolve.js:715:14)
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:808:11)
    at Loader.resolve (internal/modules/esm/loader.js:82:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:226:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:51:40)
    at link (internal/modules/esm/module_job.js:50:36)
```

and import is ` import ' fs?commonjs-require';` (there's invisible NUL in the beginning?)

without the workaroundn it fails with

```
(node:9089) UnhandledPromiseRejectionWarning: ReferenceError: __nollup__external__fs__default__ is not defined
    at Object.2 (file:///home/ivk/dev/nollup-bug/build/util1-[hash].js:38:30)
    at create_module (file:///home/ivk/dev/nollup-bug/build/index.js:76:32)
    at file:///home/ivk/dev/nollup-bug/build/index.js:78:25
    at Object.1 (file:///home/ivk/dev/nollup-bug/build/util1-[hash].js:34:19)
    at create_module (file:///home/ivk/dev/nollup-bug/build/index.js:76:32)
    at _require (file:///home/ivk/dev/nollup-bug/build/index.js:192:21)
    at cb (file:///home/ivk/dev/nollup-bug/build/index.js:155:37)
```

which seems to be that other problem of missing imports

**Comment by charlag** [comment]
This NUL is  '\0' and is convention for "virtual modules" it seems
> If your plugin uses 'virtual modules' (e.g. for helper functions), prefix the module ID with \0. This prevents other plugins from trying to process it.

**Comment by charlag** [comment]
Here it takes the original import, not what it resolves to:

https://github.com/PepsRyuu/nollup/blob/b99dcd340c58dc551674d72cc569958a421e2b6d/lib/impl/ImportExportResolver.js#L195-L197

**Comment by charlag** [comment]
I propose this:

```js
dependency.source = resolved && resolved.id || node.source.value;
```

but can be more precise

**Comment by charlag** [comment]
This seems to be not enough as it still yields empty modules for some reason...

**Comment by charlag** [comment]
This is the problem only for modules like `export {default} from "tty";` which `plugin-commonjs` apparently emits.
Trying to figure where everything disappears, probably it's a special ast node type...

**Comment by charlag** [comment]
The problem is that import and export is the same node and in the end of `resolveImport` the node is overwritten

**Comment by PepsRyuu** [comment]
Highly appreciate all of the research. Few things been occupying my time but I promise I'll get to look into this soon. :)

**Comment by charlag** [comment]
Don't worry, I really appreciate that you react. Take your time. I fixed the issue with export from in our fork but run into another one caused by missing live bindings I think but we can live with that.

**Comment by PepsRyuu** [comment]
Just a comment to say that I am looking into this now. Hopefully will have a fix soon! :)

**Comment by PepsRyuu** [comment]
https://github.com/PepsRyuu/nollup/pull/172/files

It seems to work with your reproduction. :)

Still need to figure out how to test this properly, but would appreciate it if you could give the branch a try!

**Comment by PepsRyuu** [comment]
Test cases written, will merge and release tomorrow unless something else comes up!

**Comment by charlag** [comment]
Hey thanks a lot, will try today, for me it worked w/ this branch:
https://github.com/tutao/nollup/commits/dynamic-import-ext

**Comment by PepsRyuu** [comment]
Released in ```0.14.4```

**Comment by charlag** [comment]
Thanks, I didn't make it this year, will try it out when I'll get back to work.
Happy holidays!

**Comment by charlag** [comment]
I'm sorry but it seems like the bug is still there. We have a reproducer:

https://github.com/ganthern/nollup-bug/tree/synthetic-crypto

**Comment by PepsRyuu** [comment]
Thanks, will try get back to you on this as soon as I can. :)

**Comment by PepsRyuu** [comment]
Looks like the fix accidentally got deleted during the large refactor I did for ```0.15.0```. 😳 Need to look into those test cases to make sure it doesn't happen again.

**Comment by PepsRyuu** [comment]
Fixed in ```0.15.1```. Test cases should cover it now so shouldn't happen again. Apologies. :)

**Comment by charlag** [comment]
Thanks a lot!
